cpp/unreal_plugins/UrdfBot/Source/UrdfParser.cpp

```diff
@@ -0,0 +1,428 @@
+//
+// Copyright(c) 2022 Intel. Licensed under the MIT License <http://opensource.org/licenses/MIT>.
+//
+
+#include "UrdfParser.h"
+
+#include <XmlFile.h>
+#include <XmlNode.h>
+#include <XmlParser.h>
+
+#include "Assert/Assert.h"
+
+UrdfRobotDesc UrdfParser::parse(const std::string& file_name)
+{
+    FXmlFile file;
+    file.LoadFile(UTF8_TO_TCHAR(file_name.c_str()));
+
+    FXmlNode* robot_node = file.GetRootNode();
+    ASSERT(robot_node->GetTag().Equals(TEXT("robot")));
+
+    return parseRobotNode(robot_node);
+}
+
+UrdfGeometryDesc UrdfParser::parseGeometryNode(FXmlNode* geometry_node)
+{
+    UrdfGeometryDesc geometry;
+
+    const TArray<FXmlNode*>& child_nodes = geometry_node->GetChildrenNodes();
+    ASSERT(child_nodes.Num() == 1);
+    FXmlNode* node = child_nodes[0];
+    const FString& tag = node->GetTag();
+
+    if (tag.Equals(TEXT("box"))) {
+        geometry.type_ = UrdfGeometryType::Box;
+        geometry.size_ = parseVector(node->GetAttribute("size"));
+    } else if (tag.Equals(TEXT("cylinder"))) {
+        geometry.type_ = UrdfGeometryType::Cylinder;
+        geometry.length_ = FCString::Atof(*node->GetAttribute("length"));
+        geometry.radius_ = FCString::Atof(*node->GetAttribute("radius"));
+    } else if (tag.Equals(TEXT("sphere"))) {
+        geometry.type_ = UrdfGeometryType::Sphere;
+        geometry.radius_ = FCString::Atof(*node->GetAttribute("radius"));
+    } else if (tag.Equals(TEXT("mesh"))) {
+        geometry.type_ = UrdfGeometryType::Mesh;
+        geometry.filename_ = TCHAR_TO_UTF8(*node->GetAttribute("filename"));
+        FString scale = node->GetAttribute("scale");
+        if (!scale.IsEmpty()) {
+            geometry.scale_ = FCString::Atof(*scale);
+        }
+    } else {
+        ASSERT(false);
+    }
+
+    return geometry;
+}
+
+UrdfMaterialDesc UrdfParser::parseMaterialNode(FXmlNode* material_node)
+{
+    UrdfMaterialDesc material_desc;
+
+    material_desc.name_ = TCHAR_TO_UTF8(*material_node->GetAttribute(TEXT("name")));
+    material_desc.is_reference_ = true; // override default value
```

Comment:
- `// override default value` comment is not needed (technically this comment applies to all assignments)

Reply:
Material node is not required, and it will be `false` if there is not one in xml. Having no material node is meanful (use mesh's default material and do not override it) and we want a simple way to check it. Having default value of `true` with empty name do indicate it is default value though. I just want to point it out so maybe we could have better design here.

Reply:
Ah you're right. Never mind. We should default to `false` because this line of code doesn't execute unconditionally. It only executes when a material node is present.